### PR TITLE
Create NEST Extension Interface for NEST compatible with new registration scheme and unloading

### LINKED ISSUE #3064: Issues with new model registration in NEST
Following #2990, the model registration and the semantics of `ResetKernel()` have changed. After incorporating these changes to NESTML models, the tests fail when there are multiple calls to the `ResetKernel()`. For example,

```python
nest.ResetKernel()
nest.Install("nestmlmodule")

nrn = nest.Create(“neuron1”)
...
nest.Simulate(100.)
...
nest.ResetKernel()  # removes the neuron1 and neuron2 from the list of registered models
# nest.Install("nestmlmodule”)   # Error: Module 'nestmlmodule' is loaded already.
nrn = nest.Create(“neuron2”)  # Neuron cannot be created - Unknown neuron model
```

In the above example `nestmlmodule` is an external module consisting of models `neuron1` and `neuron2`. According to the new semantics of `ResetKernel()`, a call to the function deregisters the models loaded from an external module. Due to this, the second call to `ResetKernel()` de-registers `neuron1` and `neuron2` which were previously installed from the `nest.Install()` call. Hence, the `neuron2` cannot be created and the script fails. 

A second call to `nest.Install()` after the second `ResetKernel()` also fails as the same module cannot be loaded again. This is a known error as the external modules are still wrapped as SLI modules and can't be unloaded/reloaded.

As a result of these issues, the NESTML master breaks with the latest NEST master. Therefore, for the new model registration to work with NESTML, the module unloading/reloading needs to be in place.

### PR DESCRIPTION
**Pull request by heplesser**
The title says it all. This is still in draft stage as I need to iron out some problems, especially segfaults on `lt_dlclose()` of the module. `ResetKernel()` works and one can install the module again afterwards, but I have not yet checked what happens if the module changes between reloads. At present, I need to provide the `(DY)LD_LIBRARY_PATH` to the install location of the module. The old `DynamicLoader` is replaced by `ModuleManager` and `NESTExtensionInterface` is the base class for extension modules.

In https://github.com/nest/nest-extension-module/pull/26 you will find a version of `mymodule` (in the `modulemgr` branch) that works with this branch of NEST.

This fixes #3064.

### CODE REVIEW

**Comment by heplesser** [comment]
Everything works now, including loading a modified module after `ResetKernel()`. One still needs to make sure that the path to the extension module is explicitly set in `(DY)LD_LIBRARY_PATH`, even if it is the same as the installation location for the NEST libs. Got to see if one can improve on this, would need some search-path expansion in the module manager. But otherwise, this is ready for review.

File: CMakeLists.txt [hunk 1]
```diff
@@ -67,7 +67,6 @@ set( with-modelset "full" CACHE STRING "The modelset to include. Sample configur
 set( with-models OFF CACHE STRING "The models to include as a semicolon-separated list of model headers (without the .h extension). This option is mutually exclusive with -Dwith-modelset. [default=OFF]." )
 set( tics_per_ms "1000.0" CACHE STRING "Specify elementary unit of time [default=1000 tics per ms]." )
 set( tics_per_step "100" CACHE STRING "Specify resolution [default=100 tics per step]." )
-set( external-modules OFF CACHE STRING "External NEST modules to be linked in, separated by ';', [default=OFF]." )
```
Comment (pnbabu) [inline CMakeLists.txt:70]: Could you also update this in the docs?

**Review by pnbabu — CHANGES_REQUESTED** [review]
Thank you for these changes! I tested this with modules generated from NESTML and the install works fine. I have a few comments. Please find them below.

File: nestkernel/module_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    std::string msg = "Module '" + name + "' could not be opened.";
+    if ( errstr )
+    {
+      msg += "\nThe dynamic loader returned the following error: '" + std::string( errstr ) + "'.";
+    }
+    msg += "\n\nPlease check LD_LIBRARY_PATH (OSX: DYLD_LIBRARY_PATH)!";
+    throw DynamicModuleManagementError( msg );
+  }
+
+  // see if we can find the register_components symbol in the module
```
[the remark this answers is not in the document]
Comment (heplesser) [reply nestkernel/module_manager.cpp:131]: Yes, thanks for catching this.

File: nestkernel/nest_extension_interface.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  virtual std::string name() = 0;
+
+  //! Contact information
+  virtual std::string contact() = 0;
+
+  //! Website
+  virtual std::string url() = 0;
+
+  //! Permanent URI
+  virtual std::string uri() = 0;
```
[the remark this answers is not in the document]
Comment (heplesser) [reply nestkernel/nest_extension_interface.h:78]: In principle yes, but getting that right will require a bit more thought, so I would leave it for a later PR.

File: CMakeLists.txt (hunk 1 above)
Comment (heplesser) [reply CMakeLists.txt:70]: Fixed.

**Comment by heplesser** [comment]
@pnbabu Thank you for your review! I have fixed most items and commented on the two remaining.

File: nestkernel/module_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+ModuleManager::install( const std::string& name )
+{
+  if ( name.empty() )
+  {
+    throw DynamicModuleManagementError( "Module name must not be empty." );
+  }
+
+  if ( modules_.find( name ) != modules_.end() )
+  {
+    throw DynamicModuleManagementError( "Module '" + name + "' is loaded already." );
```
[the remark this answers is not in the document]
Comment (pnbabu) [reply nestkernel/module_manager.cpp:110]: You are right, reloading a module without `ResetKernel()` can lead to chaos if there is a network created with an older version of the module. 

> Thinking of it, I wonder if we should not prohibit module loading after any network components have been created and also after Simulate() has been called (in case someone calls it on an empty network).

Shouldn't the call to `Install()` be prohibited if there are some network components already created?

[the remark this answers is not in the document]
Comment (pnbabu) [reply nestkernel/module_manager.cpp:110]: @heplesser ping!

[the remark this answers is not in the document]
Comment (heplesser) [reply nestkernel/module_manager.cpp:119]: @clinssen Actually, I am not sure what state one would end up in if we load the same dynamic module twice without unloading it first. The new version of the same name might even contain different code. It seems much safer to prohibit this.

@pnbabu I have now prohibited Install if nodes exist.

The `name`  in the checks here is actually the file name. But I agree that the `name()` field is a bit double up, so we could leave it out for now and then introduce metadata when we know what we want.

[the remark this answers is not in the document]
Comment (clinssen) [reply nestkernel/module_manager.cpp:119]: I'm fine with disallowing loading a module by the same name twice. In that case, I would suggest to work purely on the basis of the filename and leave out the ``name`` member string in the extension module interface for now.

**Comment by heplesser** [comment]
@clinssen  This PR should definitely go into 3.7, but after discussion with @terhorstd I remove the milestone again for technical reasons.

**Comment by heplesser** [comment]
@clinssen @pnbabu I have now removed the four metadata methods from the `NESTExtensionModule` interface. I also made a corresponding update to the extension module branch.

**Review by pnbabu — APPROVED** [review]
Looks good to me, thank you!

**Comment by heplesser** [comment]
Marked as blocked, because I just discovered a bug in the PR: Changing the number of threads will unload all models from extension modules. I know how to fix this. Also switching to Draft to avoid it being merged by accident.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:70]: Static linking is intentionally removed because it was complex to implement and as far as I know little used. We can re-instate it later if there should be actual demand.

[the remark this answers is not in the document]
Comment (terhorstd) [reply CMakeLists.txt:70]: How do you come to the insight "as far as I know little used"? – Would it make sense to at least improve documentation on how to build modules with a notice that static modules are no longer supported and how users could change to dynamic modules?

File: cmake/ProcessOptions.cmake [hunk 5]
```diff
@@ -304,7 +267,7 @@ function( NEST_PROCESS_WITH_READLINE )
   if ( with-readline )
     if ( NOT ${with-readline} STREQUAL "ON" )
       # a path is set
-      set( READLINE_ROOT_DIR "${with-readline}" )
```
Comment (terhorstd) [inline cmake/ProcessOptions.cmake:307]: I don't see why are these changes actually necessary. Did something in CMake itself require these changes?

File: nestkernel/kernel_manager.h [hunk 6]
```diff
@@ -263,9 +264,18 @@ class KernelManager
    */
   void write_to_dump( const std::string& msg );
 
+  /**
+   * \defgroup Manager components in NEST kernel
+   *
+   * The managers are defined below in the order in which they need to be initialized.
```
Comment (terhorstd) [inline nestkernel/kernel_manager.h:270]: The order of initialization is defined by the initializer list in `kernel_manager.cpp`, so this statement is a bit misleading. I'd try to rephrase:
```suggestion
   * The list managers below should be kept in the same order in which they are
     initialized in `kernel_manager.cpp`.
```

File: nestkernel/module_manager.h [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+  void
+  install( const std::string& name )
+  {
+    throw KernelException( "Dynamic modules not supported without libltdl." );
+  }
+};
+}
+
+#endif
```
Comment (terhorstd) [inline nestkernel/module_manager.h:131]:
```suggestion
#endif  // HAVE_LIBLTDL
```

File: nestkernel/module_manager.h [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+// Includes from nestkernel:
+#include "manager_interface.h"
+
+// Includes from sli:
+#include "dictutils.h"
+
+// DynamicLoaderModule defined only if libltdl is available
+#ifdef HAVE_LIBLTDL
+#include <ltdl.h>
```
Comment (terhorstd) [inline nestkernel/module_manager.h:41]: Does it make sense to process any `#includes` if there is no LTDL?
```suggestion
// DynamicLoaderModule defined only if libltdl is available
#ifdef HAVE_LIBLTDL

#define MODULE_MANAGER_H

// Generated includes:
#include "config.h"

// C++ includes:
#include <map>
#include <string>

// Includes from nestkernel:
#include "manager_interface.h"

// Includes from sli:
#include "dictutils.h"

#include <ltdl.h>
```

**Review by terhorstd — CHANGES_REQUESTED** [review]
Some minor points and some questions still open.

File: nestkernel/module_manager.h (hunk 7 above)
Comment (pnbabu) [inline nestkernel/module_manager.h:131]:
```suggestion
#endif // #ifdef HAVE_LIBLTDL
```

File: nestkernel/module_manager.h [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  install( const std::string& name )
+  {
+    throw KernelException( "Dynamic modules not supported without libltdl." );
+  }
+};
+}
+
+#endif
+
+#endif
```
Comment (pnbabu) [inline nestkernel/module_manager.h:133]:
```suggestion
#endif // #ifndef MODULE_MANAGER_H
```

**Review by pnbabu — APPROVED** [review]
Looks good to me! Just some minor suggestions.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (heplesser) [reply CMakeLists.txt:70]: There are currently only 12 forks of the extension module (EM) repository, of which only four by people outside the core team, so the EM mechanism is little used outside of NESTML. We changed the interface with NEST 3.6 (new model registration scheme) and there has been no questions/comments/complaints about this since December. The static linking mechanism is mentioned in NEST only in the CMake flags documentation (I will check that I removed it there) and as final item in the Extension Module documentation (https://nest-extension-module.readthedocs.io/en/latest/extension_modules.html#linking-mymodule-into-nest). The process is rather cumbersome, so if it had been used, I am sure there would have been questions. But I did not find a single mail in the NEST User archives concerning the -Dexternal-modules switch.

File: nestkernel/nestmodule.cpp [hunk 10]
```diff
@@ -94,7 +94,7 @@ NestModule::~NestModule()
 const std::string
 NestModule::name() const
 {
-  return std::string( "NEST Kernel 2" ); // Return name of the module
+  return std::string( "NEST Kernel 3" ); // Return name of the module
```
[the remark this answers is not in the document]
Comment (heplesser) [reply nestkernel/nestmodule.cpp:97]: I assume the numbers reflects the NEST major version, which we forgot to change when we created NEST 3.0. So I figured better late than never. But we could also drop the number entirely and just return `"NEST Kernel"`. This name is only shown when starting NEST with verbosity `M_DEBUG` or below or in error messages is `nestmodule` cannot be loaded on startup.

File: cmake/ProcessOptions.cmake (hunk 5 above)
Comment (heplesser) [reply cmake/ProcessOptions.cmake:307]: During debugging work on LDTL loading, I found that our `-Dwith-ldtl` switch did not allow us to give a specific ltdl as argument, because we were setting the wrong variables: They need to be `<Packagename>_ROOT` (see https://cmake.org/cmake/help/latest/envvar/PackageName_ROOT.html) and newer CMAKEs (since 3.27) handle also uppercased package names (see https://cmake.org/cmake/help/latest/policy/CMP0144.html#policy:CMP0144)

File: nestkernel/kernel_manager.h (hunk 6 above)
Comment (heplesser) [reply nestkernel/kernel_manager.h:270]: Members are initialized in order of declaration in the class definition, not in the order of the initializer list in the constructor. Compilers (should) warn about it if the order in the initializer list differs from order of declaration. The reason for this is that the destructor needs to destruct in reverse order of initialization and that is only reliably known from the declaration order; see section "Initialization order" in https://en.cppreference.com/w/cpp/language/constructor.

File: nestkernel/module_manager.h (hunk 8 above)
Comment (heplesser) [reply nestkernel/module_manager.h:41]: We need `config.h` to get `HAVE_LIBLTDL`, `dictutils.h` and `manager_interface.h` for the definition of the empty class when LTDL is missing, and we need to define `MODULE_MANAGE_H` in any case to avoid double inclusion of the header. So the only difference would be `map` and `string`,  and there I think it keeps the code simpler as is. The reduction in compile time due to the reduced includes would very limited.

File: nestkernel/module_manager.h (hunk 7 above)
Comment (heplesser) [reply nestkernel/module_manager.h]: Fixed by accepting @terhorstd's equivalent suggestion.

File: nestkernel/nestmodule.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (heplesser) [reply nestkernel/nestmodule.cpp]: I removed the version number now.

**Comment by heplesser** [comment]
@terhorstd Thanks for your comments! Could you take a look at my replies?

File: nestkernel/random_manager.cpp [hunk 11]
```diff
@@ -171,8 +171,8 @@ nest::RandomManager::set_status( const DictionaryDatum& d )
 
   if ( rng_seed_updated or rng_type_updated )
   {
-    finalize( /* reset_kernel */ false );
-    initialize( /* reset_kernel */ false );
+    finalize( /* adjust_number_of_threads_only */ true );
```
Comment (clinssen) [inline nestkernel/random_manager.cpp:174]:
```suggestion
    finalize( /* adjust_number_of_threads_or_rng_seed_only */ true );
```
debates about named parameters to function calls aside (maybe in a future C++ standard?), should this parameter name be consistent along managers, or did you name it in more specific ways for specific managers?

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (clinssen) [reply CMakeLists.txt:70]: Just for posterity, what is the argument for removing this functionality? For example, could statically linked module names conflict with dynamically loaded names, as we check whether modules are already loaded on the basis of a list of dynamic module names which doesn't include the static ones?

File: .#foo [hunk 12]
```diff
@@ -0,0 +1 @@
+plesser@Hanss-MacBook-Pro.local.9953
```
[the remark this answers is not in the document]
Comment (heplesser) [reply .#foo]: Thanks for catching this, removed now.

**Review by clinssen — APPROVED** [review]

**Review by terhorstd — APPROVED** [review]
Looks good so far. Thanks for the complex work!

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (clinssen) [reply CMakeLists.txt:70]: As per the 2024 Q1 NEST hackathon, we decided to remove support for statically linked extension modules for now. The management of potential naming conflicts, and adding the requisite support in combination with the new loading/unloading mechanism would be prohibitive. Furthermore, we are not aware of this mechanism being used in practice in the past (many) years. Please write to us on the NEST Simulator mailing list if you require this functionality and we can discuss it further.
